mypy/metastore.py

```diff
@@ -0,0 +1,172 @@
+"""Interfaces for accessing metadata.
+
+We provide two implementations.
+ * The "classic" file system implementation, which uses a directory
+   structure of files.
+ * A hokey sqlite backed implementation, which basically simulates
+   the file system in an effort to work around poor file system performance
+   on OS X.
+"""
+
+import binascii
+import sqlite3
+import os
+import time
+
+from abc import abstractmethod
+from typing import Dict, List, Set, Iterable, Any, Optional
+
+
+class MetadataStore:
+    @abstractmethod
+    def getmtime(self, name: str) -> float: ...
+
+    @abstractmethod
+    def read(self, name: str) -> str: ...
+
+    @abstractmethod
+    def write(self, name: str, data: str, mtime: Optional[float] = None) -> bool: ...
+
+    @abstractmethod
+    def remove(self, name: str) -> None: ...
+
+    @abstractmethod
+    def commit(self) -> None:
+        """If the backing store requires a commit, do it.
+
+        But N.B. that this is not *guarenteed* to do anything, just to be necessary
+        with the sqlite backend.
+        """
+        pass
+
+    @abstractmethod
+    def list_all(self) -> Iterable[str]: ...
+
+
+def random_string() -> str:
+    return binascii.hexlify(os.urandom(8)).decode('ascii')
+
+
+class FilesystemMetadataStore(MetadataStore):
+    def __init__(self, cache_dir_prefix: str) -> None:
+        self.cache_dir_prefix = cache_dir_prefix
+
+    def getmtime(self, name: str) -> float:
+        return int(os.path.getmtime(os.path.join(self.cache_dir_prefix, name)))
+
+    def read(self, name: str) -> str:
+        assert os.path.normpath(name) != os.path.abspath(name), "Don't use absolute paths!"
+
+        with open(os.path.join(self.cache_dir_prefix, name), 'r') as f:
+            return f.read()
+
+    def write(self, name: str, data: str, mtime: Optional[float] = None) -> bool:
+        assert os.path.normpath(name) != os.path.abspath(name), "Don't use absolute paths!"
+
+        path = os.path.join(self.cache_dir_prefix, name)
+        tmp_filename = path + '.' + random_string()
+        try:
+            os.makedirs(os.path.dirname(path), exist_ok=True)
+            with open(tmp_filename, 'w') as f:
+                f.write(data)
+                f.write('\n')
```

Comment:
This produces an easily observable difference between the two cache classes: if you write() then read(), this class will append an extra \n while the sqlite class doesn't. Maybe just stop appending the \n? (I was going to suggest naming the method writeln, but that's not really appropriate since we don't want to do this for sqlite.)

Reply:
Yeah I will just ditch it.